### DISCUSSION TITLE
CMS support

### DISCUSSION THREAD

**Discussion by MichaelAuckland** [statiqdev/Discussions#71] [Q&A]
Do you have any info on how to integrate Statiq with a headless CMS? There is a module which can fetch data from Kentico CMS and feed it into Statiq. However, I am looking to use some open source headless CMS instead. Has this been done before?

**Comment by daveaglick** [comment]
Great question! There's an old module for Wyam that talks to Contentful - it hasn't been ported yet: https://github.com/contentful/contentful.wyam. I don't know of any others yet besides Kentico and Contentful though. The both use different conventions and ways of inserting data though, so both serve as interesting examples.

In general, this is exactly the kind of scenario Statiq is trying to solve as a "framework" - providing enough tools and flexibility that pulling data from anywhere, whether it's a headless CMS or a database, becomes easy enough to integrate into a static site workflow.

The level of integration probably changes depending on if you're looking at Statiq Web or a fully custom Statiq-based site using Statiq Framework. I'll assume Statiq Web for now. There are probably a few possible styles of getting CMS data and providing to Statiq Web depending on if you want the resulting content to be processed like files on the disk, if they come across fully rendered, or some combination of the two.

In general though you want to write (or add) modules that get the data from the headless CMS and provide them to the other existing pipelines. The `Execute` module is particularly helpful for one-off code inside a module. Right now this all requires a little poking around with how pipelines and modules work and the dependencies between pipelines in Statiq Web.

One of the first things I want to start working on after finishing up Statiq Docs though is native support for CMS systems in Statiq Web. I'd love to have a consistent extensibility mechanism for services or other "document providers". I think that would make plugging in a headless CMS or any other system a bit easier in Statiq Web.

Did this make sense? It might have been a bit rambling. I'm happy to help with any specific questions - or I might be able to put together a quick little example of how it could work structurally using fake data.
